parlai/chat_service/core/chat_service_manager.py

```diff
@@ -0,0 +1,677 @@
+            )
+            overworld_fut.cancel()
+
+        # 5 minute grace period for conversations to finish
+        time_passed = 0
+        while time_passed < 5 * 60:
+            any_alive = False
+            for task_fut in self.active_worlds.values():
+                if task_fut is not None:
+                    any_alive = any_alive or task_fut.running()
+            if not any_alive:
+                break
+            time.sleep(1)
+
+        # Tell the worlds that they should be shutting down at this point,
+        # agents will refer to this value
+        self.shutting_down = True
+
+    def _get_unique_pool(self):
+        """Return unique pool.
+
+        Returns a filtered version of the agent pool where each agent is
+        only listed a maximum of one time.
+
+        :return:
+            a dictionary mapping world_types to agent pools
+        """
+        valid_pools = {}
+        for world_type, agent_pool in self.agent_pool.items():
+            eligibility_function = shared_utils.get_eligibility_fn(
+                self.world_module, world_type
+            )
+            if eligibility_function is not None:
+                valid_pools[world_type] = [
+                    w for w in agent_pool if eligibility_function(w)
+                ]
+            else:
+                valid_pools[world_type] = self.agent_pool[world_type]
+        return valid_pools
+
+    @abstractmethod
+    def restructure_message():
+        """Use this function to restructure the message into the provided format."""
+
+    @abstractmethod
+    def _handle_bot_read(self, agent_id):
+        """Use this function to handle/execute events once the bot has observed the message."""
+
+    @abstractmethod
+    def _confirm_message_delivery(self, event):
+        """A callback for when messages are marked as delivered"""
+
+    def _handle_message_read(self, event):
+        # If the message was sent by another user (as in during a conversation)
+        # then we need to propogate the read back to that user.
+        reader = event['sender']['id']
+        agent_state = self.get_agent_state(reader)
+        if agent_state is None:
+            return
+        agent = agent_state.get_active_agent()
+        if agent is not None:
+            for partner in agent.message_partners:
+                # We don't know who sent the message that was seen, but we can
+                # send a message observed event to everyone else in the chat
+                self.message_sender.send_read(partner.id)
+
+    def _on_first_message(self, message):
+        """Handle first message from player.
+
+        Run when a psid is given that is not paired with any assignment yet.
+        Launch an overworld, complete onboarding, etc.
+
+        :param message:
+            message sent from agent
+        """
+        if self.service_reference_id is None:
+            self.service_reference_id = message['recipient']['id']
```

Comment:
is this message assuming the format defined in the README above? if so, seems that this might be different from what is going on

Reply:
Yes. I created the message format to account for that. recipient is a dict that contains only the id as an item in it. Maybe it wasn't clear since all of that was in one line in the README. Will update it to:
```config
recipient: {
    id: #id of message recipient
  }
  sender: {
    id: #id of message sender
  }
```